torch/_subclasses/fake_impls.py

```diff
@@ -764,6 +764,74 @@ def convert_tensor(t, device):
     )
 
 
+@register_op_impl(aten._scaled_dot_product_cudnn_attention.default)
+def meta__scaled_dot_product_cudnn(fake_mode, func, *args, **kwargs):
+    _, kwargs = normalize_function(
+        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+    )
+
+    query = kwargs["query"]
+    key = kwargs["key"]
+    value = kwargs["value"]
+    compute_log_sumexp = kwargs["compute_log_sumexp"]
+    # unused: attn_bias, dropout_p, is_causal, scale
+
+    def convert_tensor(t, device):
+        return FakeTensor(fake_mode, t, device)
+
+    B = query.size(0)
+    H = query.size(1)
+    S_Q = query.size(2)
+    S_KV = key.size(2)
+    D_QK = query.size(-1)
+    D_V = value.size(-1)
+
+    res = convert_tensor(
+        torch.empty(B, H, S_Q, D_V, dtype=query.dtype, device="meta"),
+        query.device,
+    )
+
+    logsum_exp = convert_tensor(
+        torch.empty(
+            (B, H, S_Q),
+            dtype=torch.float,
+            device="meta",
+        ),
+        query.device,
+    )
+
+    # See Note [Seed and Offset]:
+    seed = convert_tensor(
+        torch.empty((), dtype=torch.long, device="meta"), query.device
+    )
+    offset = convert_tensor(
+        torch.empty((), dtype=torch.long, device="meta"), query.device
+    )
+    # Why does it cause a segfault if we replace these with None
+    # as is done in the SDPA Flash case above?
+    cum_seq_q = convert_tensor(
+        torch.empty((), dtype=torch.long, device="meta"), query.device
+    )
+    cum_seq_kv = convert_tensor(
+        torch.empty((), dtype=torch.long, device="meta"), query.device
+    )
+    debug_mask = convert_tensor(
+        torch.empty(0, dtype=query.dtype, device="meta"),
+        query.device,
+    )
+    return (
+        res,
+        logsum_exp,
+        cum_seq_q,
```

Comment:
CC'ing @davidberard98 for a question here: it looks like if I make `cum_seq_q` and `cum_seq_k` `None` here as was done in the FA version above I get a segfault in `test/inductor/test_aot_inductor.py -k test_sdpa_abi_compatible_cuda` and `test/inductor/test_aot_inductor.py -k test_sdpa_2_abi_compatible_cuda`? Is this expected? I'm just blindly pattern-matching here so it's not obvious to me what is different between these cases (e.g., the arguments do not seem to be optional in the FA version either). Note that these arguments are currently unused in the cuDNN case, but we wanted to add them to reduce the number of potential BC-breaking interface changes.

Reply:
cc @desertfire - do you know?

I do see some custom shim handling in shim_common.cpp, maybe this is where it's handled?

https://github.com/pytorch/pytorch/blob/c017c97333dfb9d17f2e5357980241827e50e8d5/torch/csrc/inductor/aoti_torch/shim_common.cpp#L469-L475

Reply:
Thanks! that's very helpful as it looks like the autogenerated shim doesn't do this 
```
        *ret0 = new_tensor_handle(std::move(std::get<0>(tmp_result)));;
        *ret1 = new_tensor_handle(std::move(std::get<1>(tmp_result)));;
        *ret2 = new_tensor_handle(std::move(std::get<2>(tmp_result)));;
        *ret3 = new_tensor_handle(std::move(std::get<3>(tmp_result)));;
        *ret4 = std::get<4>(tmp_result).expect_int();
        *ret5 = std::get<5>(tmp_result).expect_int();
        *ret6 = new_tensor_handle(std::move(std::get<6>(tmp_result)));;
        *ret7 = new_tensor_handle(std::move(std::get<7>(tmp_result)));;
        *ret8 = new_tensor_handle(std::move(std::get<8>(tmp_result)));;
```
while the autogenerated one for efficient attention had `nullptr` checks...
which led to this magic list:
https://github.com/pytorch/pytorch/blob/f7eae279463b719c5f25587aac225bd2be891373/torchgen/gen_aoti_c_shim.py#L220